[ENH] Implemented OIDC flow that results in a query sent to the f-API

rmanaem opened the pull request:
<!--- Until this PR is ready for review, you can include [WIP] in the title, or create a draft PR. -->


<!---
Below is a suggested pull request template. Feel free to add more details you feel are relevant/necessary.

For more info on the Neurobagel PR process and other contributing guidelines, see https://neurobagel.org/contributing/CONTRIBUTING/.
-->

<!-- 
Please indicate after the # which issue you're closing with this PR, if applicable.
If the PR closes multiple issues, include "closes" before each one is listed.
You can also link to other issues if necessary, e.g. "See also #1234".

https://help.github.com/articles/closing-issues-using-keywords
-->
- Closes #197 

<!-- 
Please give a brief overview of what has changed or been added in the PR.
This can include anything specific the maintainers should be looking for when they review the PR.
-->
Changes proposed in this pull request:

-
-

<!-- To be checked off by reviewers -->
## Checklist
_This section is for the PR reviewer_

- [x] PR has an interpretable title with a prefix (`[ENH]`, `[FIX]`, `[REF]`, `[TST]`, `[CI]`, `[MNT]`, `[INF]`, `[MODEL]`, `[DOC]`) _(see our [Contributing Guidelines](https://neurobagel.org/contributing/CONTRIBUTING#pull-request-guidelines) for more info)_
- [x] PR has a label for the release changelog or `skip-release` (to be applied by maintainers only)
- [x] PR links to GitHub issue with mention `Closes #XXXX`
- [x] Tests pass
- [x] Checks pass

For new features:
- [x] Tests have been added

For bug fixes:
- [ ] There is at least one test that would fail under the original bug conditions.

rmanaem reviewed (approved):
🧑‍🍳

rmanaem committed:
Renamed `NBDialog` component to `GetDataDialog`

rmanaem committed:
Added `react-oauth/google` and `jwt-decode` to dependencies

rmanaem committed:
Wrapped the app with `GoogleOAuthProvider`

rmanaem committed:
Updated import statement

rmanaem committed:
Implemented `AuthDialog` component

rmanaem committed:
Implemented Avatar for user profile pic

rmanaem committed:
Updated request config to include ID token

rmanaem committed:
Implemented action to dismiss notificaitons

rmanaem committed:
Implemented profile menu

rmanaem committed:
Implemented logout functionality

rmanaem committed:
Small clean ups

rmanaem committed:
Implemented `NB_ENABLE_AUTH` and `NB_QUERY_CLIENT_ID` env vars

rmanaem committed:
Set a conditional to render auth related components and elements

rmanaem committed:
Updated README.md

rmanaem committed:
Changed the default for `NB_ENABLE_AUTH` from `true` to `false`

rmanaem committed:
Implemented check for `NB_QUERY_CLIENT_ID`

rmanaem committed:
Fixed typo in README.md

rmanaem committed:
Implemented component test for `AuthDialog`

rmanaem committed:
Merge branch 'main' into feat-197

rmanaem committed:
Added missing props to `Navbar` component test

surchs committed:
Merge branch 'main' into feat-197

surchs commented on cypress/component/AuthDialog.cy.tsx:
```diff
@@ -0,0 +1,26 @@
+describe('ContinuousField', () => {
+  it('Displays a MUI dialog with the title and "sing in with google" button', () => {
+    cy.mount(
+      <GoogleOAuthProvider clientId="mock-client-id">
+        {' '}
+        <AuthDialog isLoggedIn={props.isLoggedIn} onAuth={props.onAuth} />
+      </GoogleOAuthProvider>
+    );
```
Is this a special trick to get the auth provider to run inside the e2e test? Shouldn't hold up the PR, but we should note that down in our notes

surchs commented on cypress/component/AuthDialog.cy.tsx:
```diff
@@ -0,0 +1,26 @@
+      'Please login using one of the following before proceeding'
+    );
+    cy.get('[data-cy="auth-dialog"]').within(() => {
+      cy.contains('Sign in with Google');
```
Pretty interesting error I ran into here: The "Sign in with Google" uses the language of the country from where you make the request. So the test would fail in Germany because here the string will be 
![image](https://github.com/user-attachments/assets/b76cc3c7-5d61-41f7-888b-0f6597adda9e)

Something to keep in mind for e2e tests. For now I'll switch the check to just "Google"

surchs committed:
[FIX] auth-component test failed in Germany

Interesting behaviour for the oauth prompt:
it is written in the local language from where the request is sent.
So from Germany it's in German.

I changed the string check and also the login prompt language

surchs commented on src/App.tsx:
```diff
@@ -308,19 +338,40 @@ function App() {
     setLoading(false);
   }
 
+  function login(credential: string | undefined) {
+    setIsLoggedIn(true);
+    const jwt: GoogleJWT = credential ? jwtDecode(credential) : ({} as GoogleJWT);
+    setIDToken(credential);
+    setName(jwt.given_name);
+    setProfilePic(jwt.picture);
+  }
```
That'd be a good candidate for a custom hook in the future afaict - so we can subscribe to the global "logged in" status with all components that need it.

surchs commented on src/components/Navbar.tsx:
```diff
@@ -20,6 +40,13 @@ function Navbar() {
       });
   }, []);
 
+  const handleClick = (event: React.MouseEvent<HTMLElement>) => {
+    setAnchorEl(event.currentTarget);
+  };
+  const handleClose = () => {
+    setAnchorEl(null);
+  };
```
We should revisit this bit soon and see if by refactoring the whole login section into its own component we can get rid of the custom click handling. Could you open an issue please.

surchs commented on src/main.tsx:
```diff
@@ -12,11 +14,15 @@ const router = createBrowserRouter([
     </StyledEngineProvider>
   </React.StrictMode>
 );
+
+ReactDOM.createRoot(document.getElementById('root')!).render(
+  enableAuth ? <GoogleOAuthProvider clientId={clientID}>{app}</GoogleOAuthProvider> : app
+);
```
neat!

surchs commented on src/App.tsx:
```diff
@@ -42,6 +47,11 @@ function App() {
   const [imagingModality, setImagingModality] = useState<FieldInput>(null);
   const [searchParams, setSearchParams] = useSearchParams();
 
+  const [isLoggedIn, setIsLoggedIn] = useState<boolean>(false);
+  const [name, setName] = useState<string>('');
+  const [profilePic, setProfilePic] = useState<string>('');
+  const [IDToken, setIDToken] = useState<string | undefined>('');
+
```
could be good candidate for single state object and reducer

surchs commented on src/App.tsx:
```diff
@@ -53,6 +63,16 @@ function App() {
       }
     : null;
 
+  const action = (snackbarId: SnackbarKey) => (
+    <IconButton
+      onClick={() => {
+        closeSnackbar(snackbarId);
+      }}
+    >
+      <CloseIcon className="text-white" />
+    </IconButton>
+  );
+
```
nice - we probably want to also refactor this soon into it's own component though

surchs commented on src/App.tsx:
```diff
@@ -308,19 +338,40 @@ function App() {
+    setName(jwt.given_name);
+    setProfilePic(jwt.picture);
+  }
+
+  function logout() {
+    googleLogout();
+    setIsLoggedIn(false);
+    setIDToken('');
+    setName('');
+    setProfilePic('');
+  }
+
```
I think that's also a good reducer candidate for a cleanup / refactor

surchs commented on src/App.tsx:
```diff
@@ -308,19 +338,40 @@ function App() {
 
+  function login(credential: string | undefined) {
+    setIsLoggedIn(true);
+    const jwt: GoogleJWT = credential ? jwtDecode(credential) : ({} as GoogleJWT);
+    setIDToken(credential);
+    setName(jwt.given_name);
+    setProfilePic(jwt.picture);
```
really cool!

surchs commented on src/App.tsx:
```diff
@@ -280,19 +302,27 @@ function App() {
     const url: string = constructQueryURL();
     try {
-      const response = await axios.get(url);
+      const response = await axios.get(url, {
+        headers: {
+          Authorization: `Bearer ${IDToken}`,
+          'Content-Type': 'application/json',
+        },
+      });
       // TODO: remove this branch once there is no more non-federation option
```
If auth is NOT enabled, does this just send the header anyway, but with an empty ID token? That's probably alright for now, but we should revisit!

surchs reviewed (approved):
Very cool PR @rmanaem! :tada: Very exciting new functionality!

I read through it and it looks and feels really nice. As you said, there are some things we can refactor / clean up soon, but the logic is really clear and the functionality works cleanly for me. I left some comments here so we can go back and look at them together. But for now I will merge this!

:cook:
